tests/unit_tests/test_time_series_any_time_series.py

```diff
@@ -0,0 +1,293 @@
+    # act
+    result = any_time_series.index_for_timestamp_not_later_than('2016-01-01T00:00:00.588Z')
+    # assert
+    assert result == 1
+
+
+@pytest.mark.parametrize('timestamps', [['2014-01-01T00:00:00.588Z', '2015-02-02T00:00:00.588Z']])
+def test_index_for_timestamp_not_later_than_same_timestamp(timestamps, example_model_and_crs):
+    # arrange
+    model, crs = example_model_and_crs
+    any_time_series = TimeSeries(model)
+    any_time_series.timestamps = timestamps
+    # act
+    result = any_time_series.index_for_timestamp_not_later_than('2015-02-02T00:00:00.588Z')
+    # assert
+    assert result == 1
+
+
+@pytest.mark.parametrize('timestamps', [['2014-01-01T00:00:00.588Z', '2015-02-02T00:00:00.588Z']])
+def test_index_for_timestamp_not_later_is_none(timestamps, example_model_and_crs):
+    # arrange
+    model, crs = example_model_and_crs
+    any_time_series = TimeSeries(model)
+    any_time_series.timestamps = timestamps
+    # act
+    result = any_time_series.index_for_timestamp_not_later_than('2013-02-02T00:00:00.588Z')
+    # assert
+    assert result is None
+
+
+@pytest.mark.parametrize('timestamps', [['2014-01-01T00:00:00.588Z', '2015-02-02T00:00:00.588Z']])
+def test_index_for_timestamp_not_earlier_than_earlier_timestamp(timestamps, example_model_and_crs):
+    # arrange
+    model, crs = example_model_and_crs
+    any_time_series = TimeSeries(model)
+    any_time_series.timestamps = timestamps
+    # act
+    result = any_time_series.index_for_timestamp_not_earlier_than('2013-02-02T00:00:00.588Z')
+    # assert
+    assert result == 0
+
+
+@pytest.mark.parametrize('timestamps', [['2014-01-01T00:00:00.588Z', '2015-02-02T00:00:00.588Z']])
+def test_index_for_timestamp_not_earlier_than_later_timestamp(timestamps, example_model_and_crs):
+    # arrange
+    model, crs = example_model_and_crs
+    any_time_series = TimeSeries(model)
+    any_time_series.timestamps = timestamps
+    # act
+    result = any_time_series.index_for_timestamp_not_earlier_than('2016-02-02T00:00:00.588Z')
+    # assert
+    assert result is None
+
+
+@pytest.mark.parametrize('timestamps', [['2014-01-01T00:00:00.588Z', '2015-02-02T00:00:00.588Z']])
+def test_index_for_timestamp_not_earlier_than_same_timestamp(timestamps, example_model_and_crs):
+    # arrange
+    model, crs = example_model_and_crs
+    any_time_series = TimeSeries(model)
+    any_time_series.timestamps = timestamps
+    # act
+    result = any_time_series.index_for_timestamp_not_earlier_than('2015-02-02T00:00:00.588Z')
+    # assert
+    assert result == 1
+
+
+@pytest.mark.parametrize('timestamps', [['2014-01-01T00:00:00.588Z', '2015-02-02T00:00:00.588Z']])
+def test_index_for_timestamp_closest_to_empty_timestamps(timestamps, example_model_and_crs):
+    # arrange
+    model, crs = example_model_and_crs
+    any_time_series = TimeSeries(model)
+    any_time_series.timestamps = None
+    # act
+    result = any_time_series.index_for_timestamp_closest_to('2000-01-01T00:00:00.588Z')
```

Comment:
Do we expect this to return None when passed a valid timestamp outside the range of the TimeSeries timestamps?  If so, I feel the docstring for the function should make that clear.  Otherwise I would expect this call to return an index of 0.

Reply:
Ah, I see, my mistake – the method returns None if and only if there are no timestamps in the series.